gpytorch/lazy/keops_lazy_tensor.py

```diff
@@ -0,0 +1,108 @@
+import torch
+from .lazy_tensor import LazyTensor
+from ..utils.memoize import cached
+from ..utils.getitem import _noop_index
+
+
+class KeOpsLazyTensor(LazyTensor):
+    def __init__(self, x1, x2, covar_func, **params):
+        super().__init__(x1, x2, covar_func=covar_func, **params)
+
+        self.x1 = x1
+        self.x2 = x2
+        self.covar_func = covar_func
+        self.params = params
+
+    @cached(name="kernel_diag")
+    def diag(self):
+        """
+        Getting the diagonal of a kernel can be handled more efficiently by
+        transposing the batch and data dimension before calling the kernel.
+        Implementing it this way allows us to compute predictions more efficiently
+        in cases where only the variances are required.
+        """
+        return self.covar_func(self.x1, self.x2, diag=True)
+
+    @property
+    @cached(name="covar_mat")
+    def covar_mat(self):
+        return self.covar_func(self.x1, self.x2, **self.params)
+
+    def _matmul(self, rhs):
+        # TODO: replace with `self.covar_mat @ rhs` on next PyKeOps release.
+        batch_shape = self.batch_shape
+        if len(batch_shape):
+            # Equivalent to dim=-3, but KeOps LT doesn't support relative index in sum.
+            return (self.covar_mat * rhs[..., None, :, :].contiguous()).sum(dim=len(batch_shape) + 1)
+        else:
+            if not rhs.is_contiguous():
```

Comment:
Nit: this is not necessary - `mat.contiguous` returns `mat` if `mat.is_contiguous`.